internal/provider/resource_user_alias.go

```diff
@@ -0,0 +1,243 @@
+package googleworkspace
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	admin "google.golang.org/api/admin/directory/v1"
+)
+
+func resourceUserAlias() *schema.Resource {
+	return &schema.Resource{
+		Description:   "User alias resource manages individual aliases for the given Google workspace account.",
+		CreateContext: resourceUserAliasCreate,
+		ReadContext:   resourceUserAliasRead,
+		UpdateContext: nil,
+		DeleteContext: resourceUserAliasDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceUserAliasImport,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+		},
+		Schema: map[string]*schema.Schema{
+			"primary_email": {
+				Type:        schema.TypeString,
+				Description: "Primary Email (userKey) of the user the alias should be applied to.",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"alias": {
+				Type:        schema.TypeString,
+				Description: "Email alias which should be applied to the user.",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"etag": {
+				Description: "ETag of the resource.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceUserAliasCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	client := meta.(*apiClient)
+
+	directoryService, diags := client.NewDirectoryService()
+	if diags.HasError() {
+		return diags
+	}
+
+	usersService, diags := GetUsersService(directoryService)
+	if diags.HasError() {
+		return diags
+	}
+
+	aliasesService, diags := GetUserAliasService(usersService)
+	if diags.HasError() {
+		return diags
+	}
+
+	primaryEmail := d.Get("primary_email").(string)
+	setAlias := d.Get("alias").(string)
+
+	alias := &admin.Alias{
+		Alias: setAlias,
+	}
+	alias, err := aliasesService.Insert(primaryEmail, alias).Do()
+	if err != nil {
+		return diag.Errorf("[ERROR] failed to add alias for user (%s): %v", primaryEmail, err)
+	}
+
+	// Using a different backoff style due to the fact the top level ETag is not a consistent value for individual aliases. In theory it will change when multiple
```

Comment:
This is interesting! I was unfamiliar with it. I've struggled a lot with the eventual consistency of these resources, so this is a nice new tool to have available. Thank you!

Reply:
Cheers! Yea I learned this pattern from the okta provider, since they also have eventual consistency gotchas, but no immediate feedback in the form of an etag or similar pattern.

I personally prefer this pattern since I think the ETag tracking and counting can get pretty hairy, considering gotchas like the one I mentioned here. For me doing this plus just waiting for your change to be read out correctly (like the read function would do) is the least complex solution.